### ISSUE TITLE
Multiple batches created due to asyncio scheduler and locking

### ISSUE THREAD

**Issue by tyoung-patreon** [strawberry-graphql/strawberry#2186]
## Description

I encountered a problem where the batching mechanism of dataloaders isn't working due to the asyncio scheduler not guaranteeing execution order. I've provided some example code in a pytest below demonstrating the issue. Broadly what is happening is that a top level fan out request all go through some asyncio locked mechanism and then proceed to hit a dataloader. They all hit the lock and are blocked, and then proceed to the dataloader. However asyncio doesn't execute all of the locked tasks before proceeding with the dataloader execution in small batches of 2.

```py
from typing import List

import asyncio
import dataclasses
import pytest
from strawberry.dataloader import DataLoader


def test_batching_with_caching():
    batch_invocations = 0

    lock = asyncio.Lock()
    result = None

    @dataclasses.dataclass(frozen=True)
    class Result:
        x: int = 0

    async def load_result(keys: List[int]) -> List[Result]:
        nonlocal batch_invocations
        batch_invocations += 1
        print(f"> invoking batch < {keys}")
        return [Result(x=key) for key in keys]

    loader = DataLoader(load_fn=load_result)

    async def cached_result(i: int):
        nonlocal lock, result
        print(f"locking..: {i}")
        async with lock:
            print(f"acquired lock: {i}")
            # Trivial example of a task that needs to be behind a lock
            result = await asyncio.sleep(0)  # yield to scheduler
            print(f"done with locked task.. {i}")
        print(f"released lock: {i}")
        return result

    async def example_fn(i: int):
        await cached_result(i)
        return await loader.load(i)

    coros = list(example_fn(i) for i in range(10))
    asyncio.get_event_loop().run_until_complete(asyncio.gather(*coros))
    assert batch_invocations == 1
```

This results in 5 batches instead of 1. Debugging information output:

```
locking..: 0
acquired lock: 0
locking..: 1
locking..: 2
locking..: 3
locking..: 4
locking..: 5
locking..: 6
locking..: 7
locking..: 8
locking..: 9
done with locked task.. 0
released lock: 0
acquired lock: 1
done with locked task.. 1
released lock: 1
> invoking batch < [0, 1]
acquired lock: 2
done with locked task.. 2
released lock: 2
acquired lock: 3
done with locked task.. 3
released lock: 3
> invoking batch < [2, 3]
acquired lock: 4
done with locked task.. 4
released lock: 4
acquired lock: 5
done with locked task.. 5
released lock: 5
> invoking batch < [4, 5]
acquired lock: 6
done with locked task.. 6
released lock: 6
acquired lock: 7
done with locked task.. 7
released lock: 7
> invoking batch < [6, 7]
acquired lock: 8
done with locked task.. 8
released lock: 8
acquired lock: 9
done with locked task.. 9
released lock: 9
> invoking batch < [8, 9]
```

This doesn't seem like it is the dataloaders fault but instead with how the asyncio scheduler is switching between tasks which it knows are waiting for a lock and those which are not. Is there anyway I can get around this?

**Comment by patrick91** [comment]
hi @tyoung-patreon, I'll check this in the next few days, but I was wondering if you tried this while running a GraphQL query 🤔

**Comment by tyoung-patreon** [comment]
Thanks for the detailed response! It is taking me a while to internalize what is going on.

@patrick91  - I am not using graphql yet, it is a slow migration

@jkimbo - I'm not sure I understand the assertion:

>  await loader.load(i) only gets called sequentially after after each cached_result has finished`

I don't think that is the case, because while each of the `cached_result` calls must happen sequentially, each task yields to the scheduler when invoking the dataloader, which could all still happen in one large batch. Here is modified example that uses `asyncio.sleep(1)` where only the first task performs the yield and all the others return a cached value.


```py
from typing import List

import asyncio
import dataclasses
import pytest
from strawberry.dataloader import DataLoader


def test_batching_with_caching():
    batch_invocations = 0

    lock = asyncio.Lock()
    done = False
    result = None

    @dataclasses.dataclass(frozen=True)
    class Result:
        x: int = 0

    async def load_result(keys: List[int]) -> List[Result]:
        nonlocal batch_invocations
        batch_invocations += 1
        print(f"> invoking batch < {keys}")
        return [Result(x=key) for key in keys]

    loader = DataLoader(load_fn=load_result)

    async def cached_result(i: int):
        nonlocal lock, result, done
        print(f"locking..: {i}")
        async with lock:
            print(f"acquired lock: {i}")
            if not done:
                print(f"Executing.. {i}")
                result = await asyncio.sleep(1) # yield to scheduler
                print(f"Done.. {i}")
                done = True
        print(f"released lock: {i}")
        return result

    async def example_fn(i: int):
        await cached_result(i)
        x = await loader.load(i)
        print(f"Finished load {i}")
        return x

    coros = list(example_fn(i) for i in range(10))
    asyncio.get_event_loop().run_until_complete(asyncio.gather(*coros))
    assert batch_invocations == 1
 ```

The output of this is batches of size 3:

```
locking..: 0
acquired lock: 0
Executing.. 0
locking..: 1
locking..: 2
locking..: 3
locking..: 4
locking..: 5
locking..: 6
locking..: 7
locking..: 8
locking..: 9
Done.. 0
released lock: 0
acquired lock: 1
released lock: 1
acquired lock: 2
released lock: 2
> invoking batch < [0, 1, 2]
acquired lock: 3
released lock: 3
Finished load 0
Finished load 1
Finished load 2
acquired lock: 4
released lock: 4
acquired lock: 5
released lock: 5
> invoking batch < [3, 4, 5]
acquired lock: 6
released lock: 6
Finished load 3
Finished load 4
Finished load 5
acquired lock: 7
released lock: 7
acquired lock: 8
released lock: 8
> invoking batch < [6, 7, 8]
acquired lock: 9
released lock: 9
Finished load 6
Finished load 7
Finished load 8
> invoking batch < [9]
Finished load 9
```

I do not understand why this would be the case unless the scheduler is treating tasks blocked on a lock differently than those that are not. The execution order is:

1. Task 0 gets the lock and yields
2. Tasks 1-9 are blocked on the lock
3. Task 0 releases lock and continues, yielding on dataloader
4. Task 1 resumes and yields on dataloader
5. Task 2 resumes and yields on dataloader
6. Scheduler now runs dataloader batched task instead of task 3 (which has been waiting longer)
7. Task 3 resumes and yields on dataloader
8. Task 0 1 and 2 run
9 ... patern continues


Step 6 is really what is confusing me. Why is the dataloader running instead of task 3, which has been enqueued for longer and the lock is free?

**Comment by jkimbo** [comment]
@tyoung-patreon that execution order you outlined looks correct and I'm also seeing the batching of 3 that you are. I'm not sure exactly why it's batching in 3s but it must be something to do with how the asyncio scheduler manages things. You can see in the dataloader source that all we're doing is calling `call_soon` on the loop to schedule the batching: https://github.com/strawberry-graphql/strawberry/blob/9475dc6d6d41049bca3440f187c4d49ac37363bb/strawberry/dataloader.py#L182 (ref: https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_soon)

It's possible that we could expose a way of configuring that so that you could use `call_later` with a non zero value to make sure that all `dataloader.load` calls have been made before executing the batch. But I'm not sure it's a good idea because you'd be always adding in a fixed delay into your API calls.

**Comment by tyoung-patreon** [comment]
Yeah I figured it wasn't something caused by the dataloader code base, I just wanted to figure out how to manage these types of patterns at scale. I believe if everything is through dataloaders it works fairly well, but when you start mixing in other things developers might do with asyncio and really deep call hierarchies it seems things can start to go wrong. I noticed this because it looked like we had an "n plus 1" query when we were using dataloaders, and it took a fair bit of time to debug and figure out this is what was happening.

Another logical example instead of a lock would be each task doing something that takes a variable amount of time and does not go through the dataloader api. This could be simulated with `asyncio.sleep(randint(0,5))`.

In these example I'm not sure what the "logically correct" thing to even do is. I agree that you don't want to just needlessly add delay to the dataloader tasks, but it seems that it shouldn't eagerly schedule either. It seems to me that if there were some prioritization mechanism in the asyncio scheduler we could take advantage of, it would be useful to schedule batch tasks as low priority as possible (so that more callers have the opportunity to join the batch). I'm unaware of whether such a mechanism exists, and I also understand that isn't necessarily desirable for all users.

For now I've gotten around my issue, but I would still love to hear your thoughts around best practices for using dataloaders at scale in fairly deep and wide call hierarchies.
